controllers/btpoperator_controller_test.go

```diff
@@ -544,11 +531,6 @@ var _ = Describe("BTP Operator controller", Ordered, func() {
 			err := k8sClient.Delete(ctx, configMap)
 			Expect(err).To(BeNil())
 		})
-
-		AfterAll(func() {
-			//graceful shutdown
-			time.Sleep(time.Second * 1)
```

Comment:
removing this sleep causes error in AfterSuite node:
```
[AfterSuite] 
.../btp-manager/controllers/suite_test.go:125
  STEP: tearing down the test environment @ 01/13/23 11:07:50.297
  2023-01-13T11:07:50.297+0100	INFO	Stopping and waiting for non leader election runnables
  2023-01-13T11:07:50.297+0100	INFO	Stopping and waiting for leader election runnables
  2023-01-13T11:07:50.297+0100	INFO	Shutdown signal received, waiting for all workers to finish	{"controller": "btpoperator", "controllerGroup": "operator.kyma-project.io", "controllerKind": "BtpOperator"}
  2023-01-13T11:07:50.300+0100	ERROR	while doing ConsistencyCheck	{"controller": "btpoperator", "controllerGroup": "operator.kyma-project.io", "controllerKind": "BtpOperator", "btpOperator": {"name":"btp-operator-test","namespace":"default"}, "namespace": "default", "name": "btp-operator-test", "reconcileID": "550cbcd2-389c-4bc4-9d3b-18fea90dff6b", "error": "could not get information about the resource: Get \"https://127.0.0.1:51886/apis/rbac.authorization.k8s.io/v1/clusterroles/sap-btp-operator-metrics-reader\": dial tcp 127.0.0.1:51886: connect: connection refused", "errorVerbose": "Get \"https://127.0.0.1:51886/apis/rbac.authorization.k8s.io/v1/clusterroles/sap-btp-operator-metrics-reader\": dial tcp 127.0.0.1:51886: connect: connection refused\ncould not get information about the resource\nhelm.sh/helm/v3/pkg/kube.(*Client).Update.func1\n\t.../go/pkg/mod/helm.sh/helm/v3@v3.10.3/pkg/kube/client.go:238\nhelm.sh/helm/v3/pkg/kube.ResourceList.Visit\n\t.../go/pkg/mod/helm.sh/helm/v3@v3.10.3/pkg/kube/resource.go:32\nhelm.sh/helm/v3/pkg/kube.(*Client).Update\n\t.../go/pkg/mod/helm.sh/helm/v3@v3.10.3/pkg/kube/client.go:230\ngithub.com/kyma-project/module-manager/pkg/manifest.(*helm).installResources\n\t.../go/pkg/mod/github.com/kyma-project/module-manager@v0.0.0-20221214222100-348f6d59ed0b/pkg/manifest/helm_processor.go:339\ngithub.com/kyma-project/module-manager/pkg/manifest.(*helm).IsConsistent\n\t.../go/pkg/mod/github.com/kyma-project/module-manager@v0.0.0-20221214222100-348f6d59ed0b/pkg/manifest/helm_processor.go:286\ngithub.com/kyma-project/module-manager/pkg/manifest.(*Operations).consistencyCheck\n\t.../go/pkg/mod/github.com/kyma-project/module-manager@v0.0.0-20221214222100-348f6d59ed0b/pkg/manifest/operations.go:228\ngithub.com/kyma-project/module-manager/pkg/manifest.ConsistencyCheck\n\t.../go/pkg/mod/github.com/kyma-project/module-manager@v0.0.0-20221214222100-348f6d59ed0b/pkg/manifest/operations.go:71\ngithub.com/kyma-project/btp-manager/controllers.(*BtpOperatorReconciler).chartConsistencyCheck\n\t.../btp-manager/controllers/btpoperator_controller.go:1170\ngithub.com/kyma-project/btp-manager/controllers.(*BtpOperatorReconciler).HandleReadyState\n\t.../btp-manager/controllers/btpoperator_controller.go:1191\ngithub.com/kyma-project/btp-manager/controllers.(*BtpOperatorReconciler).Reconcile\n\t.../btp-manager/controllers/btpoperator_controller.go:382\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t.../go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:121\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t.../go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:320\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t.../go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t.../go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:234\nruntime.goexit\n\t/opt/homebrew/Cellar/go/1.19.4/libexec/src/runtime/asm_arm64.s:1172"}
  github.com/kyma-project/btp-manager/controllers.(*BtpOperatorReconciler).HandleReadyState
  	.../btp-manager/controllers/btpoperator_controller.go:1191
  github.com/kyma-project/btp-manager/controllers.(*BtpOperatorReconciler).Reconcile
  	.../btp-manager/controllers/btpoperator_controller.go:382
  sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
  	.../go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:121
  sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
  	.../go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:320
  sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
  	.../go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:273
  sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
  	.../go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:234
  2023-01-13T11:07:50.301+0100	ERROR	Reconciler error	{"controller": "btpoperator", "controllerGroup": "operator.kyma-project.io", "controllerKind": "BtpOperator", "btpOperator": {"name":"btp-operator-test","namespace":"default"}, "namespace": "default", "name": "btp-operator-test", "reconcileID": "550cbcd2-389c-4bc4-9d3b-18fea90dff6b", "error": "client rate limiter Wait returned an error: context canceled"}
  sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
  	.../go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:273
  sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
  	.../go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:234
  2023-01-13T11:07:50.301+0100	INFO	All workers finished	{"controller": "btpoperator", "controllerGroup": "operator.kyma-project.io", "controllerKind": "BtpOperator"}
  2023-01-13T11:07:50.301+0100	INFO	Stopping and waiting for caches
  2023-01-13T11:07:50.301+0100	INFO	Stopping and waiting for webhooks
  2023-01-13T11:07:50.301+0100	INFO	Wait completed, proceeding to shutdown the manager
[AfterSuite] PASSED [1.094 seconds]
```
We don't check for reconciler errors there, so tests are green.

Reply:
thanks, completely missed that due to test success! I will take a look if there is anything I can do to make the test more deterministic.

Reply:
proposing deterministic solution in https://github.com/kyma-project/btp-manager/pull/94/commits/bd3dc338fa00c613a3b49d97000697611095c030

Ideally I was trying to get controller workqueue [`Len()`](https://github.com/kubernetes/client-go/blob/241dfe3630210fb5c94ec73b2b8acceb72e65ea4/util/workqueue/queue.go#L141-L148), but the `controller-runtime` has the `Controller` struct under [`internal`](https://github.com/kubernetes-sigs/controller-runtime/blob/8499b67e316a03b260c73f92d0380de8cd2e97a1/pkg/internal/controller/controller.go#L62) so accessing that might be very fragile and make upgrading this module a nightmare.

Reply:
looks ok, AfterSuite finishes without error